### PR TITLE
Feat: Super users can delete discord group roles

### LINKED ISSUE #2191: Feature: Super user can delete Discord groups
### Issue Description
Super users cannot delete discord groups from the '/groups' endpoint on the dashboard site.

### Expected Behavior
At the [Discord Groups](https://dashboard.realdevsquad.com/groups/) Page, the superusers should be able to delete existing groups from the UI.

### Current Behavior
Super users don't have the option to delete discord groups through UI on the [Discord Groups](https://dashboard.realdevsquad.com/groups/) page.

### Reproducibility

 - [x] This issue is reproducible
 - [ ] This issue is not reproducible

### Steps to Reproduce
1.  Login as a super user at the Dashboard site.

2. Visit [Discord Groups](https://dashboard.realdevsquad.com/groups/).
You shall not see a delete / remove group option.

### Severity/Priority
 - [ ] Critical
 - [ ] High
 - [x] Medium
 - [ ] Low

### Checklist

- [x] I have read and followed the project's code of conduct.
- [x] I have searched for similar issues before creating this one.
- [x] I have provided all the necessary information to understand and reproduce the issue.
- [x] I am willing to contribute to the resolution of this issue.

### PR DESCRIPTION
**Pull request by VaibhavSingh8**
Date: 04 Nov 2024

Developer Name: @VaibhavSingh8

---

## Issue Ticket Number
 
- Closes #2191 

## Description

This PR implements the ability for super users to delete Discord groups through the `/groups` endpoint on the dashboard site. This feature allows for better group management.

  - Added the `deleteGroupRole` model function to handle the soft deletion of group roles in the Firestore database.
  - The deletion is implemented as a soft delete, setting `isDeleted: true` along with deletion metadata.
  - Implemented controller logic to handle deletion requests, verify role existence, log deletion events for audit, and manage responses with appropriate status codes.
  - Added comprehensive test coverage for both success and error scenarios.
  
### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [x] Yes
- [ ] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

https://github.com/user-attachments/assets/5c6c85d0-bed8-4481-af0c-78a898f40922

https://github.com/user-attachments/assets/7037cc1d-1ee5-4e62-b868-b2b9d2aef603

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

Screenshot 2024-11-20 at 2 46 54 AM

Screenshot 2024-11-20 at 2 48 24 AM

Screenshot 2024-11-20 at 2 48 48 AM

</details>

### CODE REVIEW

**Comment by VaibhavSingh8** [comment]
Hi @yesyash @vinit717 ,

Please review my PR.

Thank you!

**Comment by vinit717** [comment]
> Hi @yesyash @vinit717 ,
> 
> Please review my PR.
> 
> Thank you!

Try to be in Discord so we can track this,  ask @ankushdharkar  for further assistance

**Comment by pankajjs** [comment]
This is failing.

File: services/discordService.js [hunk 1]
```diff
@@ -103,15 +103,51 @@ const setUserDiscordNickname = async (userName, discordId) => {
     };
   } catch (err) {
     logger.error("Error in updating discord Nickname", err);
-    throw err;
+    throw new Error(err);
```
Comment (pankajjs) [inline services/discordService.js:106]: Why this change?

**Comment by pankajjs** [comment]
Test is missing for controller.

File: controllers/discordactions.js [hunk 2]
```diff
@@ -63,6 +64,79 @@ const createGroupRole = async (req, res) => {
...
+ */
+const deleteGroupRole = async (req, res) => {
+  const isDevMode = req.query.dev === "true";
+
+  if (isDevMode) {
+    try {
+      const { groupId } = req.params;
+      const { roleid } = req.body;
+
+      const { roleExists, existingRoles } = await discordRolesModel.isGroupRoleExists({ roleid });
```
[the remark this answers is not in the document]
Comment (VaibhavSingh8) [reply controllers/discordactions.js:82]: Nope, roleid is the ID of the particular group stored in DB on the creation of a group-role

[the remark this answers is not in the document]
Comment (VaibhavSingh8) [reply controllers/discordactions.js:82]: groupId, differentiates on UI which group's delete button has been clicked

File: services/discordService.js (hunk 1 above)
Comment (VaibhavSingh8) [reply services/discordService.js:106]: This is just a chore change for consistent error type and is useful in better debugging and error handling.

File: controllers/discordactions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pankajjs) [reply controllers/discordactions.js:82]: just for confirmation, groupId is id in db and roleId is id in discord.

[the remark this answers is not in the document]
Comment (VaibhavSingh8) [reply controllers/discordactions.js:82]: Exactly!

[the remark this answers is not in the document]
Comment (pankajjs) [reply controllers/discordactions.js:82]: const { roleExists, existingRoles } = await discordRolesModel.isGroupRoleExists({ roleid });
what it does?

File: controllers/discordactions.js [hunk 3]
```diff
@@ -63,6 +64,79 @@ const createGroupRole = async (req, res) => {
...
+      // To be implemented later after availability of Discord API
+
+      // const discordDeletionSuccess = await discordServices.deleteGroupRoleFromDiscord(roleid);
+
+      // if (!discordDeletionSuccess) {
+      //   console.log(`Failed to delete role from Discord with ID: ${roleid}`);
+      //   return res.status(500).json({
+      //     error: "Failed to delete role from Discord server",
+      //   });
+      // }
```
Comment (vinit717) [inline controllers/discordactions.js:101]: Remove the console log statements

File: models/discordactions.js [hunk 4]
```diff
@@ -46,6 +46,31 @@ const createNewRole = async (roleData) => {
...
+ * @param {string} groupId - The ID of the group role to be deleted
+ * @param {string} deletedBy - The ID of the user performing the deletion for logging purpose
+ * @returns {Promise<Object>} An object indicating whether the operation was successful
+ */
+const deleteGroupRole = async (groupId, deletedBy) => {
+  try {
+    const roleRef = admin.firestore().collection("discord-roles").doc(groupId);
+    await roleRef.update({
+      isDeleted: true,
+      deletedAt: admin.firestore.Timestamp.fromDate(new Date()),
```
Comment (vinit717) [inline models/discordactions.js:63]: can you please confirm it that 
admin.firestore.Timestamp.fromDate(new Date())
is adding seconds and nanoseconds or not?

File: test/unit/models/discordactions.test.js [hunk 5]
```diff
@@ -242,6 +243,66 @@ describe("discordactions", function () {
...
+      const mockFirestore = {
+        collection: () => ({
+          doc: () => ({
+            update: async () => {
+              throw new Error("Database error");
+            },
+          }),
+        }),
+      };
+      // Replacing firestore implementation with our mock
```
Comment (vinit717) [inline test/unit/models/discordactions.test.js:284]: removes the comments

**Review by vinit717 — CHANGES_REQUESTED** [review]
Add a proper test coverage and tests of the code you made

File: models/discordactions.js (hunk 4 above)
Comment (VaibhavSingh8) [reply models/discordactions.js:63]: Yes, the nanoseconds can't be seen in Firestore UI, but I confirm they are present in db. Please see the log in the screenshot attached.
Screenshot 2024-11-11 at 9 30 14 AM

File: test/unit/models/discordactions.test.js (hunk 5 above)
Comment (VaibhavSingh8) [reply test/unit/models/discordactions.test.js:284]: Sure, will remove in the changes

File: controllers/discordactions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (VaibhavSingh8) [reply controllers/discordactions.js:82]: Addressed your change of fetching roleid directly from db

File: controllers/discordactions.js (hunk 3 above)
Comment (VaibhavSingh8) [reply controllers/discordactions.js]: Removed the logs

**Comment by VaibhavSingh8** [comment]
> Add a proper test coverage and tests of the code you made

@vinit717 Can you please elaborate more on this as in what to add?

**Comment by vinit717** [comment]
> > Add a proper test coverage and tests of the code you made
> 
> @vinit717 Can you please elaborate more on this as in what to add?

Proper test coverage report of the code/file you changed

**Comment by VaibhavSingh8** [comment]
> > > Add a proper test coverage and tests of the code you made
> > 
> > 
> > @vinit717 Can you please elaborate more on this as in what to add?
> 
> Proper test coverage report of the code/file you changed

@vinit717 updated the reports in the PR description. Do I need to add tests for the controller too, because I can't find any written earlier for any controller?

**Comment by vinit717** [comment]
> updated the reports in the PR description. Do I need to add tests for the controller too, because I can't find any written earlier for any controller?

Integration test itself is controller tests make logic parts in models so you can test it

**Comment by VaibhavSingh8** [comment]
> > updated the reports in the PR description. Do I need to add tests for the controller too, because I can't find any written earlier for any controller?
> 
> Integration test itself is controller tests make logic parts in models so you can test it

Hi @vinit717 

I've implemented the requested changes. Please review the PR.

Thank you!

**Comment by pankajjs** [comment]
Are we not integrating the Discord API as we discussed? @VaibhavSingh8

**Comment by VaibhavSingh8** [comment]
> Are we not integrating the Discord API as we discussed? @VaibhavSingh8

The timeline for this issue is concise, so as I requested yesterday, let's proceed with the original advice by Ankush to do it in a separate PR.

**Comment by pankajjs** [comment]
Okay, I will pick in new task.

File: controllers/discordactions.js [hunk 6]
```diff
@@ -63,6 +64,89 @@ const createGroupRole = async (req, res) => {
...
+
+      // To be implemented later after availability of Discord API
+
+      // const discordDeletionSuccess = await discordServices.deleteGroupRoleFromDiscord(roleid);
+
+      // if (!discordDeletionSuccess) {
+      //   return res.status(500).json({
+      //     error: "Failed to delete role from Discord server",
+      //   });
+      // }
```
Comment (vinit717) [inline controllers/discordactions.js:111]: Why are we not implementing this as Pankaj has done with this?

Comment (VaibhavSingh8) [reply controllers/discordactions.js:111]: Initially, since the API was not implemented, Ankush divided this task into two parts and limited the scope of this PR to db changes only and directed the discord integration into a separate task to Pankaj.

And, due to some issues, this task has already been delayed too much that's why I was planning to go with the initial flow only, integrating in separate PR.

Suggestions are welcome.

**Comment by pankajjs** [comment]
Left few comments @VaibhavSingh8

File: controllers/discordactions.js [hunk 7]
```diff
@@ -63,6 +64,74 @@ const createGroupRole = async (req, res) => {
...
 };
 
+/**
+ * Controller function to handle the soft deletion of a group role.
+ *
+ * @param {Object} req - The request object
+ * @param {Object} res - The response object
+ * @returns {Promise<void>}
+ */
+const deleteGroupRole = async (req, res) => {
```
Comment (pankajjs) [inline controllers/discordactions.js:74]: - Refactor this controller. Place everything in separate service.
- What will happen if the delete role from discord gets successful and db query fails?

File: services/discordService.js [hunk 8]
```diff
@@ -103,7 +103,43 @@ const setUserDiscordNickname = async (userName, discordId) => {
...
+    });
+
+    if (!response.ok) {
+      return { success: false, message: "Failed to delete role from discord" };
+    }
+    if (response.status === 204) {
+      return { success: true, message: "Role deleted successfully" };
+    }
+    const data = await response.json();
+    return data;
```
Comment (pankajjs) [inline services/discordService.js:139]: The response does not contain any body as far as I know.

**Comment by pankajjs** [comment]
Left few comments @VaibhavSingh8

File: controllers/discordactions.js (hunk 7 above)
Comment (VaibhavSingh8) [reply controllers/discordactions.js:74]: Can you explain on how to make a separate service. I've followed the same pattern as in the file

File: services/discordService.js (hunk 8 above)
Comment (VaibhavSingh8) [reply services/discordService.js:139]: Made for future usage if after refactor some sends 200. Should I remove?

File: controllers/discordactions.js (hunk 7 above)
Comment (pankajjs) [reply controllers/discordactions.js:74]: - Controller should not be doing all these thing. It should accept request, call service, receive output and send response.
 - Write a function in discord-service file.

Comment (VaibhavSingh8) [reply controllers/discordactions.js:74]: There is already a function in discord service file which controller is calling

File: services/discordService.js (hunk 8 above)
Comment (pankajjs) [reply services/discordService.js:139]: yes

File: controllers/discordactions.js (hunk 7 above)
Comment (pankajjs) [reply controllers/discordactions.js:74]: We have deleteRoleFromDiscord function.
We can write another one deleteGroupRole right? and refactor controller.

Comment (VaibhavSingh8) [reply controllers/discordactions.js:74]: Not getting your point

File: services/discordService.js (hunk 8 above)
Comment (VaibhavSingh8) [reply services/discordService.js]: Done

**Comment by VaibhavSingh8** [comment]
> -In the working video, include the discord part as well. Left few comments @VaibhavSingh8

Fixed

File: controllers/discordactions.js [hunk 9]
```diff
@@ -63,6 +64,60 @@ const createGroupRole = async (req, res) => {
...
+
+    if (!discordDeletion.success) {
+      return res.boom.badImplementation(discordDeletion.message);
+    }
+
+    const { isSuccess } = await discordRolesModel.deleteGroupRole(groupId, req.userData.id);
+
+    if (!isSuccess) {
+      logger.error(`Role deleted from Discord but failed to delete from database for groupId: ${groupId}`);
+      return res.boom.badRequest("Group role deletion failed");
```
Comment (pankajjs) [inline controllers/discordactions.js:96]: why badrequest?

File: test/unit/services/discordService.test.js [hunk 10]
```diff
@@ -6,12 +6,14 @@ const {
...
   removeRoleFromUser,
   setUserDiscordNickname,
+  deleteGroupRoleFromDiscord,
 } = require("../../../services/discordService");
 const { fetchAllUsers } = require("../../../models/users");
 const Sinon = require("sinon");
 const userModel = firestore.collection("users");
 const userDataArray = require("../../fixtures/user/user")();
 const discordMembersArray = require("../../fixtures/discordResponse/discord-response");
+// const { func } = require("joi");
```
Comment (pankajjs) [inline test/unit/services/discordService.test.js:16]: remove this

File: services/discordService.js [hunk 11]
```diff
@@ -103,7 +103,40 @@ const setUserDiscordNickname = async (userName, discordId) => {
...
+ *
+ * @param {string} roleId - The Discord ID of the role to be deleted
+ * @returns {Promise<Object>} The response from the Discord API
+ * @throws {Error} If the deletion fails or there's a network error
+ */
+
+const deleteGroupRoleFromDiscord = async (roleId) => {
+  try {
+    const authToken = generateAuthTokenForCloudflare();
+    const response = await fetch(`${DISCORD_BASE_URL}/roles/${roleId}?dev=true`, {
```
Comment (pankajjs) [inline services/discordService.js:124]: create this url as constant variable and use it here

Comment (VaibhavSingh8) [reply services/discordService.js:124]: I think this is fine, take reference from other function in the file

Comment (VaibhavSingh8) [reply services/discordService.js:124]: I think this is fine, take reference from other function in the file

File: test/unit/services/discordService.test.js (hunk 10 above)
Comment (VaibhavSingh8) [reply test/unit/services/discordService.test.js:16]: I commented cuz it was not used but not sure if I can remove.

File: controllers/discordactions.js (hunk 9 above)
Comment (VaibhavSingh8) [reply controllers/discordactions.js:96]: Fixed. badImplementation

File: services/discordService.js (hunk 11 above)
Comment (pankajjs) [reply services/discordService.js:124]: No, Since we have to change this url after it is in prod. We have to modify this function again.
In case of variable, this would not be the case

Comment (VaibhavSingh8) [reply services/discordService.js:124]: Only once right? Wouldn't effect on 1 function, either way we have to remove dev flag in prod.

Comment (pankajjs) [reply services/discordService.js:124]: It is better to update a constant variable rather than a working function.

Comment (pankajjs) [reply services/discordService.js:124]: Resolved

**Review by pankajjs — APPROVED** [review]
LGTM
